src/main/java/ch/njol/skript/expressions/ExprBeaconValues.java

```diff
@@ -0,0 +1,237 @@
+							return beacon.getSecondaryEffect().getType();
+						}
+					}
+					case RANGE -> {return beacon.getEffectRange();}
+					case TIER -> {return beacon.getTier();}
+				};
+				return null;
+			}
+		});
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.REMOVE_ALL || valueType == BeaconValues.TIER) {
+			return null;
+		}
+		if (isEffect && (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)) {
+			return null;
+		}
+		return CollectionUtils.array(Object.class);
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		PotionEffectType providedEffect = null;
+		double providedRange = 0;
+		if (delta != null && delta[0] != null) {
+			if (isEffect && delta[0] instanceof PotionEffectType potionEffectType) {
+				providedEffect = potionEffectType;
+			} else if (valueType == BeaconValues.RANGE) {
+				providedRange = (double) ((long) delta[0]);
+			}
+		}
+		switch (mode) {
+			case ADD -> {
+				for (Block block : getExpr().getArray(event)) {
+					Beacon beacon = (Beacon) block.getState();
+					beacon.setEffectRange(beacon.getEffectRange() + providedRange);
+					beacon.update(true);
+				}
+			}
+			case REMOVE -> {
+				for (Block block : getExpr().getArray(event)) {
+					Beacon beacon = (Beacon) block.getState();
+					beacon.setEffectRange(Math.max(beacon.getEffectRange() - providedRange, 0));
+					beacon.update(true);
+				}
+			}
+			case DELETE, RESET -> {
+				if (valueType == BeaconValues.RANGE) {
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.resetEffectRange();
+						beacon.update(true);
+					}
+				} else if (valueType == BeaconValues.PRIMARY) {
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.setPrimaryEffect(null);
+						beacon.update(true);
+					}
+				} else if (valueType == BeaconValues.SECONDARY)	{
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.setSecondaryEffect(null);
+						beacon.update(true);
+					}
+				}
+			}
+			case SET -> {
+				if (valueType == BeaconValues.RANGE) {
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.setEffectRange(providedRange);
+						beacon.update(true);
+					}
+				} else if (valueType == BeaconValues.PRIMARY) {
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.setPrimaryEffect(providedEffect);
+						beacon.update(true);
+					}
+				} else if (valueType == BeaconValues.SECONDARY) {
+					for (Block block : getExpr().getArray(event)) {
+						Beacon beacon = (Beacon) block.getState();
+						beacon.setSecondaryEffect(providedEffect);
+						beacon.update(true);
+					}
+				}
+			}
+		}
+	}
```

Comment:
there's a lot of duplicate code that could possibly be solved by helper methods
Or at least broken up into smaller chunks